### ISSUE TITLE
Setting default slot devices in software lists prevents manual selection

### ISSUE THREAD

**Issue by cuavas** [mamedev/mame#10679]
Contrary to what’s implied by the “default”, setting slot devices in software lists prevents manual selection altogether.  I noticed this when attempting to test the DIY paddle device in 85fd7f8c45794284c7423d917795e9c96e08decb with Master System games.

Megumi Rescue (sms:megumi) specifies `<sharedfeat name="ctrl1_default" value="paddle" />`

I attempted to use the DIY paddle with an export console:
```
mamed sms1 megumi -ctrl1 diypaddle
```

However, the game starts with the simulated Sega HPD-200 paddle.  I think this behaviour is undesirable.

**Comment by 0kmg** [comment]
Yes, you can't override the default on the command line, so you have to manually change it in the UI. Generally though, the defaults are most useful when there's only one controller option and the software is unusable otherwise.

Anyway, the much worse problem with defaults is they are ignored when using the UI to pick from the software lists. Picking a software list item should swap the slotted controllers.

**Comment by enikland2** [comment]
From what I remember, the defaults also do not work with adapters, like on Game Gear using the mgear adapter, to load software using the SMS software list, and the cable adapter for the second player DE-9 port. That port will also be named "ctrl1" IIRC as it's the first instance but the default controller will not swap, what could be a good thing if player 1 & 2 controllers defaults are different, but not the case when I tried.

**Comment by cuavas** [comment]
But it makes debugging or testing difficult, and it discourages experimentation. That conflicts with MAME's selling points.

It also means it's useless when multiple peripherals are supported but the regular default is less than ideal.

It's bad enough that I'd consider ripping out the feature.

**Comment by 0kmg** [comment]
I'm 100% for making things overridable on the command line. In fact things like NTSC/PAL compatibility sharedfeat that can't be bypassed on the command line are highly highly annoying to me.

But still, having some automation for expected configurations is convenient. There are oddball examples like GI Joe/Action Force on a2600 that are for up to three players using a pair of paddles and a joystick and only that config, where it's just nice that MAME can load it that way without the user having to have advanced knowledge.

FWIW, at present *_default is only used in 9 SLs, and really only extensively in a2600, bbc_rom, and sms. I'm pretty sure everything in a2600 with defaults is only playable with those controllers and with the recorded number of controllers.

**Comment by ajrhacker** [comment]
When I first implemented the "_default" feature, it allowed command-line parameters to override the defaults, and also fell back to the configured defaults if the ones specified in the software list turned out not to exist on a particular system. These specifications were carelessly broken at some point when the options system was progressively refactored.

**Comment by ajrhacker** [comment]
42adf5d51de36560547449c6c279a8df9df7dcfe should be good enough a fix for now. Yes, I know devs have good reasons to be unhappy with MAME's slot option management in general, so of course it's not an ideal solution.

**Comment by 0kmg** [comment]
@ajrhacker Thanks for revisiting this. I have a question about 

> - Silently ignore software list-provided default options or values that don't exist in a particular system configuration

Would this cover the NES/Famicom situation where only certain clones have the expansion port? Previously adding a sharedfeat "exp_default" was a no-no because it triggered an exception on NES etc. With the fixes you've added it seems both NES _and_ Famicom ignore "exp_default". Is this the current expected behavior?

**Comment by ajrhacker** [comment]
> Would this cover the NES/Famicom situation where only certain clones have the expansion port?

That case was in fact what it was specifically intended to cover.

> With the fixes you've added it seems both NES *and* Famicom ignore "exp_default". Is this the current expected behavior?

No, that was a mistake of not testing an alternate implementation enough, and I fixed it in 742ef9cdd6266bb2665533c23a326e1370fb7a75.

**Comment by 0kmg** [comment]
Oh. It doesn't seem to be working on my end then. I added `<sharedfeat name="exp_default" value="fc_keyboard" />` to set `famibs20` in nes.xml as a test. That title requires a keyboard to boot—you get a gray screen otherwise. In 0.250 `mame famicom famibs20` will boot, but with top of tree the `exp` slot is empty in the UI which means you get the gray screen.

**Comment by cuavas** [comment]
> Anyway, the much worse problem with defaults is they are ignored when using the UI to pick from the software lists. Picking a software list item should swap the slotted controllers.

I tried a recent build, and this seems to use slot defaults from the software list, provided you have it set to skip the software part selection menu:
* Type “sms1” at system selection menu, hit return to select Master System
* Type “megumi” at software selection menu, hit return to select Megumi Rescue
* Select a BIOS, any BIOS
* Wait for boot, hit button 1
* Watch it glitch like crazy as it doesn’t have an export paddle

A 0.250 release candidate behaves the same way.
